### PR TITLE
Fixes #10195: Document server-side hooks

### PR DESCRIPTION
**Pull request by fanf**
https://www.rudder-project.org/redmine/issues/10195

### CODE REVIEW

File: 4_advanced_usage/60_server_event_hooks.txt [hunk 1]
```diff
@@ -0,0 +1,139 @@
+=== Server Event Hooks
+
+
+Rudder 4.1 introducs the possibility to execute files (hooks), typically scripts,
```
Comment (amousset) [inline 4_advanced_usage/60_server_event_hooks.txt]: introduced

File: 4_advanced_usage/60_server_event_hooks.txt [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+when some predefined event occures on Rudder.
```
Comment (amousset) [inline 4_advanced_usage/60_server_event_hooks.txt]: occurs

Comment (amousset) [inline 4_advanced_usage/60_server_event_hooks.txt]: in

File: 4_advanced_usage/60_server_event_hooks.txt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+==== Generalities about hooks
+
+*Hooks are organized in subdirectories*. The root of the sub-directories os ```/opt/rudder/etc/hooks.d/```
```
Comment (amousset) [inline 4_advanced_usage/60_server_event_hooks.txt]: is

File: 4_advanced_usage/60_server_event_hooks.txt [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+Each sub-directory has a name related to the event that will trigger the hooks
+execution.
+By default, a hook directory contains a template example and a Readme.txt file
+explaining the generalities about hooks and the specificities of that hook (parameters, etc)
```
Comment (amousset) [inline 4_advanced_usage/60_server_event_hooks.txt]: Missing . at the end

File: 4_advanced_usage/60_server_event_hooks.txt [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+*Hooks must be executable*. All executable file will be used as hooks, and
+non executable one will be ignore (which allow to put other file in these
```
Comment (amousset) [inline 4_advanced_usage/60_server_event_hooks.txt]: ignored

Comment (amousset) [inline 4_advanced_usage/60_server_event_hooks.txt]: allows to put other files

File: 4_advanced_usage/60_server_event_hooks.txt [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+directory, like a readme, for example).
```
Comment (amousset) [inline 4_advanced_usage/60_server_event_hooks.txt]: directories

File: 4_advanced_usage/60_server_event_hooks.txt [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+*Hooks parameter are passed by environment variable*. Rudder will fill dedicated
+environment variable for each hooks.
+
+*Hooks are executed sequentially, in lexical order*. We encourage
+you to use the patter "NN-hookname", with NN a number like
+"05", "20", etc.
+
+*Hooks have normalized returned code*. Return code on hooks are interpreted as follow:
```
Comment (amousset) [inline 4_advanced_usage/60_server_event_hooks.txt]: return codes

File: 4_advanced_usage/60_server_event_hooks.txt [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+- ```0     ```: success, no log (appart if debug one)          , continue to next hook
+- ```1-31  ```: error  , error   log in /var/log/rudder/webapp/, stop processing
+- ```32-63 ```: warning, warning log in /var/log/rudder/webapp/, continue to next hook
+- ```64-255```: reserved for futur use case. Behavior may change without notice.
```
Comment (amousset) [inline 4_advanced_usage/60_server_event_hooks.txt]: future

**Comment by amousset** [comment]
I don't really like copy/pasting doc from the sources, and the doc will likely become outdated at the first change in the sources.

We could extract the content from the README files (like we do for ncf generic methods or rudder command man page).

**Comment by fanf** [comment]
OK with your comment, do you know how to do that? I mean, of course technically, but more on a practical point of view, so that it is not too much fragile. Do we need some pragma/markers in the readme file ?

**Comment by amousset** [comment]
We could put a readme containing the Generalities in hooks.d directly, and avoid duplicating it in all readme files. Then we take this main readme and concatenate the readme for the different hooks folders (probably with a little tweak for  title levels). What do you think about this option?

**Comment by fanf** [comment]
seems ok

**Comment by fanf** [comment]
PR rebased

**Comment by fanf** [comment]
PR rebased

**Comment by fanf** [comment]
This pull request was superseeded by https://github.com/Normation/rudder-doc/commit/2ec4ed6052bcafa7d046dfb73f3ab77232f51a35

**Pull request closed without merging**